Adding back Observer / Live update commit as a new CR

sollipse opened the pull request:
This reverts commit f462ec2972cb14806e23e20b306c2f67734aef9c.

Commit:
Adding back Observer commit for new cr

This reverts commit f462ec2972cb14806e23e20b306c2f67734aef9c.

Commit:
Changing template

prendradjaja commented:
A couple thoughts for now: (We could find some time to pair program on this, or you could see if this is a useful enough pointer to explore on your own)

1. You can use the Observable ActivatedRoute.fragment instead of using window.onhashchange directly (this is a bit more Angular-y, and we get to delete a tiny bit of code and let the framework take care of it)
2. Instead of getting in nested promise/observable hell, we could convert cdsFetcherService.data to an Observable and then use combineLatest(? i think? i think there are a couple similar operators and i never know offhand which one is the one i want)

Commit:
Using ActivatedRoute in lieu of window.onHashChange

Converting CDSFetcherService to return an Observable instead of a promise
Converting App.Component.ts to use rxjs CombineLatest to merge
multiple subscribers.

Surpisingly simple stuff! RxJS is great.

sollipse commented:
The pointers were really helpful! combineLatest is kind of like magic, digging the multi-subscriber use pattern. The code smells much better now

prendradjaja commented on src/app/cds-fetcher.service.ts:
```diff
@@ -31,15 +32,17 @@ export type Foo = {
-  let outbreakStartIndex = dates.findIndex(([dateString, info]) => {
-    return info.cases > NUM_CASES_CUTOFF;
-  });
+  let outbreakStartIndex = dates.findIndex(
+    ([_, info]) => info.cases > NUM_CASES_CUTOFF
+  );
```
Loving the cleanup!

prendradjaja commented on src/app/cds-fetcher.service.ts:
```diff
@@ -59,17 +62,17 @@ function parseTimeseriesData(tsData: CDSLocationTimeSeriesData): Foo {
-        return response.json();
-      })
-      .then(mapValues(parseTimeseriesData));
+    this.data = new Observable((observer) => {
+      fetch(CORONA_URL)
+        // get response data to json
+        .then((res) => res.json())
+        // convert json to correctly formatted timeseries
+        .then(mapValues(parseTimeseriesData))
+        // emit the processed data.
+        .then((data) => observer.next(data));
+    });
```
Optional change: Instead of manually creating an Observable, you can use rxjs's `from()`!

(Also, if you've never seen this, it's quite handy: https://rxjs.dev/operator-decision-tree)

sollipse replied:
this is awesome

prendradjaja reviewed (approved):
Sweet! A lot cleaner. :D  One optional change

Thanks for this!

Commit:
using `from` instead of new observable

prendradjaja reviewed (approved):
✨
